### PR TITLE
Support for Console over SSH

### LINKED ISSUE #832: Feature Request: connect to device console via ssh terminal server
The netconify tool allows connections through an ssh-based terminal server that requires authentication before being presented with the device console, but it looks like PyEZ does not support this.  We do not permit unauthenticated remote access to network device consoles in our environment(and I imagine many other organizations have this requirement as well), so it would be very useful to have this functionality available in PyEZ.  I'm doing a little experimentation right now with attempting to provide a small cli tool for bootstrapping our juniper devices, and having serial/telnet functionality available directly in PyEZ but then needing to use netconify for ssh connections is not ideal.

### PR DESCRIPTION
**Pull request by mzbroch**
This feature adds the NetConf over Console over SSH transport method. Customers interested in enhanced connection security might use it with terminal servers to secure the channel. Current implementation is based on paramiko library. Tested in py2/py3 environments.

Closes #832 

cc:
@ngrundler
@jeremyschulman 
@dgarros 
@vnitinv

### CODE REVIEW

**Comment by dgarros** [comment]
@vnitinv @ydnath Please could you take a look at this PR
We have been working with @mzbroch on this enhancement, we're currently using the Console over Telnet feature and we would love to have a more secure transport.

@mzbroch PR looks good to me and I know you tested it already. 
it would be great to expose the SSH Private Key file to the paramiko library if it's provided and use it for authentication. 

Thanks

cc @jeremyschulman

File: lib/jnpr/junos/device.py [hunk 1]
```diff
@@ -1025,7 +1025,7 @@ def transform(self, func):
     # -----------------------------------------------------------------------
 
     def __new__(cls, *args, **kwargs):
-        if kwargs.get('port') in [23, '23'] or kwargs.get('mode'):
+        if kwargs.get('port') in [22, '22', 23, '23'] or kwargs.get('mode'):
```
Comment (vnitinv) [inline lib/jnpr/junos/device.py]: This condition is not correct. Even we can use port 22 to connect directly to the device (when default 830 is not configured).

File: lib/jnpr/junos/console.py [hunk 2]
```diff
@@ -274,13 +277,21 @@ def _tty_login(self):
...
+            self.console = ('ssh', self._hostname, self.port)
+            self._tty = SSH(**tty_args)
         elif self._mode.upper() == 'SERIAL':
             tty_args['port'] = self._port
             self.console = ('serial', self._port)
             self._tty = Serial(**tty_args)
         else:
-            logger.error('Mode should be either telnet or serial')
-            raise AttributeError('Mode to be telnet/serial')
+            logger.error('Mode should be one of: telnet, serial, ssh')
```
Comment (vnitinv) [inline lib/jnpr/junos/console.py]: default connection is ssh only (netconf over ssh). We cant use `ssh` for console server connection trigger point.

**Comment by vnitinv** [comment]
@mzbroch Can you please also add unit test for the same.

File: lib/jnpr/junos/console.py (hunk 2 above)
Comment (mzbroch) [reply lib/jnpr/junos/console.py]: I have renamed to cs_ssh (console server ssh).

File: lib/jnpr/junos/console.py [hunk 3]
```diff
@@ -104,6 +105,8 @@ def __init__(self, **kvargs):
             '') or kvargs.get(
             'passwd',
             '')
+        self.s_user = kvargs.get('s_user', self._auth_user)
+        self.s_passwd = kvargs.get('s_passwd', self._auth_password)
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply lib/jnpr/junos/console.py]: I have renamed into cs_user & cs_password

File: lib/jnpr/junos/console.py [hunk 4]
```diff
@@ -176,7 +179,7 @@ def open(self, *vargs, **kvargs):
         # ---------------------------------------------------------------
         # validate device hostname or IP address
         # ---------------------------------------------------------------
-        if self._mode.upper() == 'TELNET' and self._hostname is None:
+        if (self._mode.upper() == 'TELNET' or self._mode.upper() == 'SSH') and self._hostname is None:
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply lib/jnpr/junos/console.py]: I have renamed mode to 'cs_ssh' - cs_user variable has a fallback.

File: lib/jnpr/junos/console.py [hunk 5]
```diff
@@ -274,13 +277,21 @@ def _tty_login(self):
             tty_args['console_has_banner'] = self.console_has_banner
             self.console = ('telnet', self._hostname, self.port)
             self._tty = Telnet(**tty_args)
+        elif self._mode.upper() == 'SSH':
```
[the remark this answers is not in the document]
Comment (mzbroch) [reply lib/jnpr/junos/console.py]: I have renamed mode to 'cs_ssh' - cs_user variable has a fallback.

**Comment by mzbroch** [comment]
Nitin, I addressed all requested issues. Hope this works for you. I also added support for SSH keys.

**Pull request merged**